File: src/motionlab_gait/analysis/landmark_processor.py
from __future__ import annotations

import math

import numpy as np
from scipy.signal import butter, sosfiltfilt

from motionlab_gait.analysis.types import ProcessingReport
from motionlab_gait.domain.models import POSE_LANDMARK_NAMES, PoseFrame, PoseLandmark


def _finite_segments(values: np.ndarray) -> list[tuple[int, int]]:
    mask = np.isfinite(values)
    starts = np.flatnonzero(mask & ~np.r_[False, mask[:-1]])
    ends = np.flatnonzero(mask & ~np.r_[mask[1:], False]) + 1
    return list(zip(starts.tolist(), ends.tolist(), strict=True))


class LandmarkProcessor:
    def __init__(
        self,
        *,
        confidence_threshold: float = 0.5,
        max_interpolation_gap_ms: int = 100,
        cutoff_hz: float = 6.0,
        filter_order: int = 4,
    ) -> None:
        self.confidence_threshold = confidence_threshold
        self.max_interpolation_gap_ms = max_interpolation_gap_ms
        self.cutoff_hz = cutoff_hz
        self.filter_order = filter_order

    def process(
        self, frames: tuple[PoseFrame, ...], *, fps: float
    ) -> tuple[tuple[PoseFrame, ...], ProcessingReport]:
        if not frames:
            return (), ProcessingReport(
                0, 0, 0, 0, self.confidence_threshold, self.max_interpolation_gap_ms
            )

        frame_count = len(frames)
        timestamps = np.asarray([frame.timestamp_ms for frame in frames], dtype=float)
        coords = np.full((frame_count, 33, 6), np.nan, dtype=float)
        visibility = np.zeros((frame_count, 33), dtype=float)
        presence = np.zeros((frame_count, 33), dtype=float)
        originally_valid = np.zeros((frame_count, 33), dtype=bool)
        observed = np.zeros((frame_count, 33), dtype=bool)

        for frame_position, frame in enumerate(frames):
            for landmark in frame.landmarks:
                if not 0 <= landmark.index < 33:
                    continue
                index = landmark.index
                observed[frame_position, index] = True
                visibility[frame_position, index] = landmark.visibility
                presence[frame_position, index] = landmark.presence
                confident = min(landmark.visibility, landmark.presence) >= self.confidence_threshold
                originally_valid[frame_position, index] = confident
                if confident:
                    coords[frame_position, index, :3] = (landmark.x, landmark.y, landmark.z)
                    coords[frame_position, index, 3:] = (
                        np.nan if landmark.world_x is None else landmark.world_x,
                        np.nan if landmark.world_y is None else landmark.world_y,
                        np.nan if landmark.world_z is None else landmark.world_z,
                    )

        outlier_mask = np.zeros((frame_count, 33), dtype=bool)
        interpolated_mask = np.zeros((frame_count, 33), dtype=bool)
        for landmark_index in range(33):
            # Detect a 1-2 frame excursion from normalized landmark trajectories.
            # The same frame mask is then applied to every coordinate so a point
            # cannot become an anatomically inconsistent mixture of samples.
            rejected_landmark = np.zeros(frame_count, dtype=bool)
            for dimension in range(3):
                rejected_landmark |= self._reject_velocity_spikes(
                    coords[:, landmark_index, dimension], timestamps, position_floor=0.015
                )
            outlier_mask[:, landmark_index] = rejected_landmark
            coords[rejected_landmark, landmark_index, :] = np.nan
            for dimension in range(6):
                series = coords[:, landmark_index, dimension]
                filled = self._interpolate_short_gaps(series, timestamps)
                interpolated_mask[:, landmark_index] |= filled
                coords[:, landmark_index, dimension] = self._smooth(series, fps)

        output: list[PoseFrame] = []
        for frame_position, source_frame in enumerate(frames):
            landmarks: list[PoseLandmark] = []
            for index, name in enumerate(POSE_LANDMARK_NAMES):
                normalized = coords[frame_position, index, :3]
                if not np.all(np.isfinite(normalized)):
                    continue
                world = coords[frame_position, index, 3:]
                was_interpolated = bool(
                    interpolated_mask[frame_position, index]
                    or (
                        not originally_valid[frame_position, index]
                        and np.all(np.isfinite(normalized))
                    )
                )
                landmarks.append(
                    PoseLandmark(
                        index=index,
                        name=name,
                        x=float(normalized[0]),
                        y=float(normalized[1]),
                        z=float(normalized[2]),
                        visibility=float(visibility[frame_position, index]),
                        presence=float(presence[frame_position, index]),
                        world_x=float(world[0]) if math.isfinite(world[0]) else None,
                        world_y=float(world[1]) if math.isfinite(world[1]) else None,
                        world_z=float(world[2]) if math.isfinite(world[2]) else None,
                        interpolated=was_interpolated,
                        outlier_replaced=bool(outlier_mask[frame_position, index]),
                    )
                )
            output.append(
                PoseFrame(
                    frame_index=source_frame.frame_index,
                    timestamp_ms=source_frame.timestamp_ms,
                    landmarks=tuple(landmarks),
                )
            )

        report = ProcessingReport(
            frame_count=frame_count,
            interpolated_values=int(interpolated_mask.sum()),
            replaced_outliers=int(outlier_mask.sum()),
            smoothing_window_frames=0,
            confidence_threshold=self.confidence_threshold,
            max_interpolation_gap_ms=self.max_interpolation_gap_ms,
            low_confidence_landmarks=int((observed & ~originally_valid).sum()),
            filter_type="butterworth_lowpass",
            filter_order=self.filter_order,
            cutoff_hz=self.cutoff_hz,
            zero_phase=True,
        )
        return tuple(output), report

    @staticmethod
    def _reject_velocity_spikes(
        values: np.ndarray, timestamps: np.ndarray, *, position_floor: float
    ) -> np.ndarray:
        """Reject only short, high-velocity excursions that reverse immediately.

        A genuine fast movement generally continues in the same direction. A pose
        jump typically produces two unusually large velocity edges with opposite
        signs and returns close to the local trajectory within one or two frames.
        """
        rejected = np.zeros(len(values), dtype=bool)
        if len(values) < 5:
            return rejected
        dt = np.diff(timestamps) / 1000.0
        velocity = np.full(len(values) - 1, np.nan, dtype=float)
        usable = (
            np.isfinite(values[:-1])
            & np.isfinite(values[1:])
            & (dt > 0.0)
            & (dt <= 0.1)
        )
        velocity[usable] = np.diff(values)[usable] / dt[usable]
        finite_velocity = velocity[np.isfinite(velocity)]
        if len(finite_velocity) < 4:
            return rejected
        median_velocity = float(np.median(finite_velocity))
        mad = float(np.median(np.abs(finite_velocity - median_velocity)))
        robust_scale = 1.4826 * mad
        velocity_floor = position_floor / max(float(np.median(dt[dt > 0])), 1e-3)
        threshold = max(velocity_floor, 6.0 * robust_scale)

        for gap in (1, 2):
            for left in range(0, len(values) - gap - 1):
                right = left + gap + 1
                if not (math.isfinite(values[left]) and math.isfinite(values[right])):
                    continue
                first_edge = velocity[left]
                last_edge = velocity[right - 1]
                if not (math.isfinite(first_edge) and math.isfinite(last_edge)):
                    continue
                if abs(first_edge) <= threshold or abs(last_edge) <= threshold:
                    continue
                if first_edge * last_edge >= 0.0:
                    continue
                interior = np.arange(left + 1, right)
                if not np.all(np.isfinite(values[interior])):
                    continue
                expected = np.interp(
                    timestamps[interior],
                    (timestamps[left], timestamps[right]),
                    (values[left], values[right]),
                )
                if float(np.max(np.abs(values[interior] - expected))) > position_floor:
                    rejected[interior] = True
        return rejected

    def _interpolate_short_gaps(self, values: np.ndarray, timestamps: np.ndarray) -> np.ndarray:
        filled = np.zeros(len(values), dtype=bool)
        valid = np.flatnonzero(np.isfinite(values))
        for left, right in zip(valid, valid[1:], strict=False):
            if right == left + 1:
                continue
            if timestamps[right] - timestamps[left] > self.max_interpolation_gap_ms:
                continue
            positions = np.arange(left + 1, right)
            values[positions] = np.interp(
                timestamps[positions],
                (timestamps[left], timestamps[right]),
                (values[left], values[right]),
            )
            filled[positions] = True
        return filled

    def _smooth(self, values: np.ndarray, fps: float) -> np.ndarray:
        output = values.copy()
        if fps <= 0.0:
            return output
        cutoff = min(self.cutoff_hz, 0.45 * fps)
        if cutoff <= 0.0:
            return output
        sos = butter(self.filter_order, cutoff, btype="lowpass", fs=fps, output="sos")
        for start, end in _finite_segments(values):
            if end - start < 15:
                continue
            try:
                output[start:end] = sosfiltfilt(sos, values[start:end])
            except ValueError:
                # Very short finite segments are intentionally left unfiltered.
                continue
        return output

File: web/src/analysis/landmarkProcessor.ts
import { POSE_LANDMARK_NAMES, type PoseFrame } from "../domain/models";
import { inferSamplingFrequency, interpolateAt, median } from "./statistics";
import type { ProcessedPoseFrame, ProcessingReport } from "./types";

const LANDMARK_COUNT = 33;
const DIMENSION_COUNT = 6;

export interface LandmarkProcessingResult {
  frames: ProcessedPoseFrame[];
  report: ProcessingReport;
  fps: number;
}

export function effectivePresence(visibility: number, presence: number | null): number {
  // Tasks Vision Web 1.0.1 does not expose presence. Visibility is the closest
  // official signal and is used only as a documented fallback in that case.
  return presence ?? visibility;
}

export function processLandmarks(
  frames: readonly PoseFrame[],
  confidenceThreshold = 0.5,
  maxInterpolationGapMs = 100,
  cutoffHz = 6
): LandmarkProcessingResult {
  const fps = inferSamplingFrequency(frames.map((frame) => frame.timestampMs));
  const timestamps = frames.map((frame) => frame.timestampMs);
  const coords = Array.from({ length: frames.length }, () =>
    Array.from({ length: LANDMARK_COUNT }, () => Array<number>(DIMENSION_COUNT).fill(Number.NaN))
  );
  const visibility = Array.from({ length: frames.length }, () => Array<number>(LANDMARK_COUNT).fill(0));
  const presence = Array.from({ length: frames.length }, () => Array<number>(LANDMARK_COUNT).fill(0));
  const originallyValid = Array.from({ length: frames.length }, () => Array<boolean>(LANDMARK_COUNT).fill(false));
  const outlierMask = Array.from({ length: frames.length }, () => Array<boolean>(LANDMARK_COUNT).fill(false));
  const interpolatedMask = Array.from({ length: frames.length }, () => Array<boolean>(LANDMARK_COUNT).fill(false));
  let presenceFallback = false;

  frames.forEach((frame, framePosition) => {
    for (const landmark of frame.landmarks) {
      if (landmark.index < 0 || landmark.index >= LANDMARK_COUNT) continue;
      const index = landmark.index;
      const landmarkPresence = effectivePresence(landmark.visibility, landmark.presence);
      presenceFallback ||= landmark.presence === null;
      visibility[framePosition]![index] = landmark.visibility;
      presence[framePosition]![index] = landmarkPresence;
      if (Math.min(landmark.visibility, landmarkPresence) < confidenceThreshold) continue;
      originallyValid[framePosition]![index] = true;
      coords[framePosition]![index] = [
        landmark.x, landmark.y, landmark.z,
        landmark.worldX ?? Number.NaN, landmark.worldY ?? Number.NaN, landmark.worldZ ?? Number.NaN
      ];
    }
  });

  for (let landmarkIndex = 0; landmarkIndex < LANDMARK_COUNT; landmarkIndex += 1) {
    const rejected = Array<boolean>(frames.length).fill(false);
    for (let dimension = 0; dimension < 3; dimension += 1) {
      const series = coords.map((frame) => frame[landmarkIndex]![dimension]!);
      rejectVelocitySpikes(series, timestamps, 0.015).forEach((value, index) => {
        rejected[index] ||= value;
      });
    }
    rejected.forEach((value, framePosition) => {
      if (!value) return;
      outlierMask[framePosition]![landmarkIndex] = true;
      coords[framePosition]![landmarkIndex]!.fill(Number.NaN);
    });
    for (let dimension = 0; dimension < DIMENSION_COUNT; dimension += 1) {
      const series = coords.map((frame) => frame[landmarkIndex]![dimension]!);
      const filled = interpolateShortGaps(series, timestamps, maxInterpolationGapMs);
      filled.forEach((value, framePosition) => {
        interpolatedMask[framePosition]![landmarkIndex] ||= value;
      });
      const smoothed = smoothFiniteSegments(series, fps, cutoffHz);
      smoothed.forEach((value, framePosition) => {
        coords[framePosition]![landmarkIndex]![dimension] = value;
      });
    }
  }

  const processedFrames = frames.map<ProcessedPoseFrame>((source, framePosition) => ({
    frameIndex: source.frameIndex,
    timestampMs: source.timestampMs,
    landmarks: Array.from({ length: LANDMARK_COUNT }, (_, index) => index)
      .filter((index) => coords[framePosition]![index]!.slice(0, 3).every(Number.isFinite))
      .map((index) => ({
        index,
        name: POSE_LANDMARK_NAMES[index]!,
        x: coords[framePosition]![index]![0]!,
        y: coords[framePosition]![index]![1]!,
        z: coords[framePosition]![index]![2]!,
        visibility: visibility[framePosition]![index]!,
        presence: presence[framePosition]![index]!,
        worldX: finiteOrNull(coords[framePosition]![index]![3]!),
        worldY: finiteOrNull(coords[framePosition]![index]![4]!),
        worldZ: finiteOrNull(coords[framePosition]![index]![5]!),
        interpolated: interpolatedMask[framePosition]![index]!
          || (!originallyValid[framePosition]![index]! && coords[framePosition]![index]!.slice(0, 3).every(Number.isFinite)),
        outlierReplaced: outlierMask[framePosition]![index]!
      }))
  }));

  return {
    frames: processedFrames,
    fps,
    report: {
      frameCount: frames.length,
      interpolatedValues: interpolatedMask.flat().filter(Boolean).length,
      replacedOutliers: outlierMask.flat().filter(Boolean).length,
      confidenceThreshold,
      maxInterpolationGapMs,
      filterType: "butterworth_lowpass",
      filterOrder: 4,
      cutoffHz,
      zeroPhase: true,
      presenceFallback
    }
  };
}

function finiteOrNull(value: number): number | null {
  return Number.isFinite(value) ? value : null;
}

function rejectVelocitySpikes(values: readonly number[], timestamps: readonly number[], positionFloor: number): boolean[] {
  const rejected = Array<boolean>(values.length).fill(false);
  if (values.length < 5) return rejected;
  const dt = values.slice(1).map((_, index) => (timestamps[index + 1]! - timestamps[index]!) / 1000);
  const velocity = dt.map((seconds, index) => {
    const before = values[index]!;
    const after = values[index + 1]!;
    return Number.isFinite(before) && Number.isFinite(after) && seconds > 0 && seconds <= 0.1
      ? (after - before) / seconds : Number.NaN;
  });
  const finiteVelocity = velocity.filter(Number.isFinite);
  if (finiteVelocity.length < 4) return rejected;
  const medianVelocity = median(finiteVelocity);
  const robustScale = 1.4826 * median(finiteVelocity.map((value) => Math.abs(value - medianVelocity)));
  const positiveDt = dt.filter((value) => value > 0);
  const velocityFloor = positionFloor / Math.max(median(positiveDt), 0.001);
  const threshold = Math.max(velocityFloor, 6 * robustScale);
  for (const gap of [1, 2]) {
    for (let left = 0; left < values.length - gap - 1; left += 1) {
      const right = left + gap + 1;
      if (!Number.isFinite(values[left]) || !Number.isFinite(values[right])) continue;
      const firstEdge = velocity[left]!;
      const lastEdge = velocity[right - 1]!;
      if (!Number.isFinite(firstEdge) || !Number.isFinite(lastEdge)
        || Math.abs(firstEdge) <= threshold || Math.abs(lastEdge) <= threshold
        || firstEdge * lastEdge >= 0) continue;
      let maximumDeviation = 0;
      let complete = true;
      for (let position = left + 1; position < right; position += 1) {
        if (!Number.isFinite(values[position])) { complete = false; break; }
        const expected = interpolateAt(timestamps[position]!, timestamps[left]!, timestamps[right]!, values[left]!, values[right]!);
        maximumDeviation = Math.max(maximumDeviation, Math.abs(values[position]! - expected));
      }
      if (complete && maximumDeviation > positionFloor) {
        for (let position = left + 1; position < right; position += 1) rejected[position] = true;
      }
    }
  }
  return rejected;
}

function interpolateShortGaps(values: number[], timestamps: readonly number[], maximumGapMs: number): boolean[] {
  const filled = Array<boolean>(values.length).fill(false);
  const valid = values.map((value, index) => Number.isFinite(value) ? index : -1).filter((index) => index >= 0);
  for (let index = 0; index < valid.length - 1; index += 1) {
    const left = valid[index]!;
    const right = valid[index + 1]!;
    if (right === left + 1 || timestamps[right]! - timestamps[left]! > maximumGapMs) continue;
    for (let position = left + 1; position < right; position += 1) {
      values[position] = interpolateAt(timestamps[position]!, timestamps[left]!, timestamps[right]!, values[left]!, values[right]!);
      filled[position] = true;
    }
  }
  return filled;
}

interface Biquad { b0: number; b1: number; b2: number; a1: number; a2: number }

function butterworthSections(fps: number, cutoffHz: number): Biquad[] {
  const cutoff = Math.min(cutoffHz, 0.45 * fps);
  if (fps <= 0 || cutoff <= 0) return [];
  const k = Math.tan(Math.PI * cutoff / fps);
  return [0.541196100146197, 1.306562964876377].map((q) => {
    const norm = 1 / (1 + k / q + k * k);
    return {
      b0: k * k * norm,
      b1: 2 * k * k * norm,
      b2: k * k * norm,
      a1: 2 * (k * k - 1) * norm,
      a2: (1 - k / q + k * k) * norm
    };
  });
}

function filterCascade(values: readonly number[], sections: readonly Biquad[]): number[] {
  let output = [...values];
  for (const section of sections) {
    const filtered = Array<number>(output.length);
    const initial = output[0] ?? 0;
    let z1 = (1 - section.b0) * initial;
    let z2 = (section.b2 - section.a2) * initial;
    output.forEach((sample, index) => {
      const result = section.b0 * sample + z1;
      z1 = section.b1 * sample - section.a1 * result + z2;
      z2 = section.b2 * sample - section.a2 * result;
      filtered[index] = result;
    });
    output = filtered;
  }
  return output;
}

function zeroPhaseFilter(values: readonly number[], sections: readonly Biquad[]): number[] {
  const padding = Math.min(values.length - 1, 15);
  const first = values[0]!;
  const last = values.at(-1)!;
  const before = values.slice(1, padding + 1).reverse().map((value) => 2 * first - value);
  const after = values.slice(values.length - padding - 1, values.length - 1).reverse().map((value) => 2 * last - value);
  const extended = [...before, ...values, ...after];
  const forward = filterCascade(extended, sections);
  const backward = filterCascade(forward.reverse(), sections).reverse();
  return backward.slice(padding, padding + values.length);
}

function smoothFiniteSegments(values: readonly number[], fps: number, cutoffHz: number): number[] {
  const output = [...values];
  const sections = butterworthSections(fps, cutoffHz);
  if (!sections.length) return output;
  let start = 0;
  while (start < values.length) {
    while (start < values.length && !Number.isFinite(values[start])) start += 1;
    let end = start;
    while (end < values.length && Number.isFinite(values[end])) end += 1;
    if (end - start >= 15) {
      zeroPhaseFilter(values.slice(start, end), sections).forEach((value, offset) => {
        output[start + offset] = value;
      });
    }
    start = end + 1;
  }
  return output;
}

